src/View/Helper/DownloadsHelper.php

```diff
@@ -0,0 +1,121 @@
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+namespace App\View\Helper;
+
+use App\View\Helper\AppHelper;
+use App\Model\Entity\LanguageNameTrait;
+
+class DownloadsHelper extends AppHelper
+{
+    use LanguageNameTrait;
+
+    public $helpers = ['Html', 'Languages'];
+
+    const PER_LANGUAGE_DIR = '/var/www-downloads/exports/per_language';
+    const DOWNLOAD_URL = 'https://downloads.tatoeba.org/exports';
+
+    /**
+     * Get all available per-language files for the given file
+     *
+     * @param string $basename Name of the file containing all languages
+     *                         (without extension)
+     *
+     * @return array           A mapping of language code => URL for file
+     **/
+    private function availableFiles($basename) {
+        $command = 'find ' . self::PER_LANGUAGE_DIR . ' -type f ' .
+                   "-name '*$basename.tsv.bz2' -printf '%P\\n' " .
+                   '2> /dev/null';
+        $stdout = trim(shell_exec($command));
+        if (empty($stdout)) {
+            return [];
+        }
+        $paths = explode("\n", $stdout);
+
+        $map = [];
+        foreach ($paths as $path) {
+            list($code, ) = preg_split('#/#', $path);
+            $url = self::DOWNLOAD_URL . DS . 'per_language' . DS . $path;
+            $map[$code] = $url;
+        }
+        return $map;
+    }
+
+    /**
+     * Create the options for the per-language file selection on the
+     * Downloads page
+     *
+     * @param string $basename Name of the file containing all languages
+     *                         (without extension)
+     *
+     * @return array           Each item is an array with keys 'language'
+     *                         and 'url'.
+     *                         The array will always contain at least one item
+     *                         (for the all-languages file).
+     **/
+    public function createOptions($basename) {
+        $urlForAll = self::DOWNLOAD_URL . DS . $basename . '.tar.bz2';
+        $options[0] = [
+            'language' => __('All languages'),
+            'url' => $urlForAll
+        ];
+
+        $urls = $this->availableFiles($basename);
+        if (!empty($urls)) {
+            $languages = array_intersect_key(
+                $this->Languages->onlyLanguagesArray(false) +
+                ['unknown' => __x('dropdown-list', 'Unknown language')],
```

Comment:
Feel free to make `unknownLanguagesArray()` accept a parameter and use `$this->Languages->unknownLanguagesArray(false)`.

Reply:
For my very first prototype I've constructed the array myself and `unknown` was in between the other languages after localized sorting.

Then I've looked around and found the methods in the `Languages` helper. But in addition to the splitting problem, `unknowLanguagesArray()` also puts `unknown` at the beginning which I personally don't like very much (after loading the page it would show the two options `All languages` and `Only sentences in Unknown language`).

So I've used the current version which puts `unknown` at the end.

Where do you think `unknown` should be put: at the beginning, at the end or in between?

Reply:
The only place `unknownLanguagesArray()` is used is the language selector on `/sentences/show_all_in`. I think it’s reasonable to put it at the end because it’s less frequently used than other languages.